File: back-end/ShopHangTet/Controllers/AddressController.cs
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;
using ShopHangTet.Data;
using ShopHangTet.Models;
using System.Security.Claims;

namespace ShopHangTet.Controllers;

[ApiController]
[Route("api/[controller]")]
[Authorize]
public class AddressController : ControllerBase
{
    private readonly ShopHangTetDbContext _context;

    public AddressController(ShopHangTetDbContext context)
    {
        _context = context;
    }

    // GET /api/Address — Lấy tất cả địa chỉ của user đang đăng nhập
    [HttpGet]
    public async Task<IActionResult> GetMyAddresses()
    {
        var userId = User.FindFirstValue(ClaimTypes.NameIdentifier)
                  ?? User.FindFirstValue("sub")
                  ?? User.FindFirstValue("id");

        if (string.IsNullOrEmpty(userId))
            return Unauthorized(new { message = "Không xác định được người dùng." });

        var addresses = await _context.Addresses
            .Where(a => a.UserId == userId)
            .OrderByDescending(a => a.IsDefault)
            .ThenByDescending(a => a.CreatedAt)
            .ToListAsync();

        return Ok(new { Success = true, Data = addresses });
    }

    // POST /api/Address — Thêm địa chỉ mới
    [HttpPost]
    public async Task<IActionResult> AddAddress([FromBody] AddressCreateDto dto)
    {
        var userId = User.FindFirstValue(ClaimTypes.NameIdentifier)
                  ?? User.FindFirstValue("sub")
                  ?? User.FindFirstValue("id");

        if (string.IsNullOrEmpty(userId))
            return Unauthorized(new { message = "Không xác định được người dùng." });

        if (string.IsNullOrWhiteSpace(dto.ReceiverName) || string.IsNullOrWhiteSpace(dto.FullAddress))
            return BadRequest(new { message = "Tên người nhận và địa chỉ không được để trống." });

        // If marked as default, unset other defaults
        if (dto.IsDefault)
        {
            var existing = await _context.Addresses.Where(a => a.UserId == userId && a.IsDefault).ToListAsync();
            foreach (var e in existing) e.IsDefault = false;
        }

        var address = new Address
        {
            UserId = userId,
            ReceiverName = dto.ReceiverName.Trim(),
            ReceiverPhone = dto.ReceiverPhone?.Trim() ?? string.Empty,
            FullAddress = dto.FullAddress.Trim(),
            IsDefault = dto.IsDefault,
            CreatedAt = DateTime.UtcNow,
        };

        _context.Addresses.Add(address);
        await _context.SaveChangesAsync();

        return Ok(new { Success = true, Data = address });
    }

    // PUT /api/Address/{id} — Cập nhật địa chỉ
    [HttpPut("{id}")]
    public async Task<IActionResult> UpdateAddress(string id, [FromBody] AddressCreateDto dto)
    {
        var userId = User.FindFirstValue(ClaimTypes.NameIdentifier)
                  ?? User.FindFirstValue("sub")
                  ?? User.FindFirstValue("id");

        var address = await _context.Addresses.FirstOrDefaultAsync(a => a.Id == id && a.UserId == userId);
        if (address == null) return NotFound(new { message = "Không tìm thấy địa chỉ." });

        if (dto.IsDefault)
        {
            var existing = await _context.Addresses.Where(a => a.UserId == userId && a.IsDefault && a.Id != id).ToListAsync();
            foreach (var e in existing) e.IsDefault = false;
        }

        address.ReceiverName = dto.ReceiverName?.Trim() ?? address.ReceiverName;
        address.ReceiverPhone = dto.ReceiverPhone?.Trim() ?? address.ReceiverPhone;
        address.FullAddress = dto.FullAddress?.Trim() ?? address.FullAddress;
        address.IsDefault = dto.IsDefault;

        await _context.SaveChangesAsync();
        return Ok(new { Success = true, Data = address });
    }

    // DELETE /api/Address/{id} — Xoá địa chỉ
    [HttpDelete("{id}")]
    public async Task<IActionResult> DeleteAddress(string id)
    {
        var userId = User.FindFirstValue(ClaimTypes.NameIdentifier)
                  ?? User.FindFirstValue("sub")
                  ?? User.FindFirstValue("id");

        var address = await _context.Addresses.FirstOrDefaultAsync(a => a.Id == id && a.UserId == userId);
        if (address == null) return NotFound(new { message = "Không tìm thấy địa chỉ." });

        _context.Addresses.Remove(address);
        await _context.SaveChangesAsync();
        return Ok(new { Success = true, Message = "Đã xoá địa chỉ." });
    }

    // PATCH /api/Address/{id}/set-default
    [HttpPatch("{id}/set-default")]
    public async Task<IActionResult> SetDefault(string id)
    {
        var userId = User.FindFirstValue(ClaimTypes.NameIdentifier)
                  ?? User.FindFirstValue("sub")
                  ?? User.FindFirstValue("id");

        var all = await _context.Addresses.Where(a => a.UserId == userId).ToListAsync();
        var target = all.FirstOrDefault(a => a.Id == id);
        if (target == null) return NotFound(new { message = "Không tìm thấy địa chỉ." });

        foreach (var a in all) a.IsDefault = a.Id == id;
        await _context.SaveChangesAsync();

        return Ok(new { Success = true, Data = target });
    }
}

public class AddressCreateDto
{
    public string ReceiverName { get; set; } = string.Empty;
    public string? ReceiverPhone { get; set; }
    public string FullAddress { get; set; } = string.Empty;
    public bool IsDefault { get; set; } = false;
}

File: back-end-nodejs/src/controllers/addressController.js
import { Router } from 'express';
import { Address } from '../models/index.js';
import { ApiResponse } from '../dtos/index.js';
import { authenticate } from '../middlewares/index.js';

/**
 * Address Controller - Tương đương AddressController.cs
 * Routes: /api/address
 */
export function createAddressRouter() {
  const router = Router();

  // All routes require auth
  router.use(authenticate);

  function getUserId(req) {
    return req.user?.sub || req.user?.id || req.user?.userId || null;
  }

  // GET /api/address
  router.get('/', async (req, res) => {
    try {
      const userId = getUserId(req);
      if (!userId) {
        return res.status(401).json(ApiResponse.error('Không xác định được người dùng.'));
      }

      const addresses = await Address.find({ userId })
        .sort({ isDefault: -1, createdAt: -1 })
        .lean();

      return res.status(200).json({ Success: true, Data: addresses });
    } catch (error) {
      console.error('Get addresses error:', error);
      return res.status(500).json(ApiResponse.error('Internal server error'));
    }
  });

  // POST /api/address
  router.post('/', async (req, res) => {
    try {
      const userId = getUserId(req);
      if (!userId) {
        return res.status(401).json(ApiResponse.error('Không xác định được người dùng.'));
      }

      const { ReceiverName, ReceiverPhone, FullAddress, IsDefault } = req.body;

      if (!ReceiverName || !FullAddress) {
        return res
          .status(400)
          .json(ApiResponse.error('Tên người nhận và địa chỉ không được để trống.'));
      }

      if (IsDefault) {
        await Address.updateMany({ userId, isDefault: true }, { $set: { isDefault: false } });
      }

      const address = await Address.create({
        userId,
        receiverName: ReceiverName?.trim(),
        receiverPhone: ReceiverPhone?.trim() || '',
        fullAddress: FullAddress?.trim(),
        isDefault: !!IsDefault,
      });

      return res.status(200).json({ Success: true, Data: address });
    } catch (error) {
      console.error('Add address error:', error);
      return res.status(500).json(ApiResponse.error('Internal server error'));
    }
  });

  // PUT /api/address/:id
  router.put('/:id', async (req, res) => {
    try {
      const userId = getUserId(req);
      if (!userId) {
        return res.status(401).json(ApiResponse.error('Không xác định được người dùng.'));
      }

      const { ReceiverName, ReceiverPhone, FullAddress, IsDefault } = req.body;
      const address = await Address.findOne({ _id: req.params.id, userId });

      if (!address) {
        return res.status(404).json(ApiResponse.error('Không tìm thấy địa chỉ.'));
      }

      if (IsDefault) {
        await Address.updateMany({ userId, isDefault: true, _id: { $ne: address._id } }, { $set: { isDefault: false } });
      }

      if (ReceiverName !== undefined) address.receiverName = ReceiverName?.trim() || '';
      if (ReceiverPhone !== undefined) address.receiverPhone = ReceiverPhone?.trim() || '';
      if (FullAddress !== undefined) address.fullAddress = FullAddress?.trim() || '';
      if (IsDefault !== undefined) address.isDefault = !!IsDefault;

      await address.save();
      return res.status(200).json({ Success: true, Data: address });
    } catch (error) {
      console.error('Update address error:', error);
      return res.status(500).json(ApiResponse.error('Internal server error'));
    }
  });

  // DELETE /api/address/:id
  router.delete('/:id', async (req, res) => {
    try {
      const userId = getUserId(req);
      if (!userId) {
        return res.status(401).json(ApiResponse.error('Không xác định được người dùng.'));
      }

      const address = await Address.findOne({ _id: req.params.id, userId });
      if (!address) {
        return res.status(404).json(ApiResponse.error('Không tìm thấy địa chỉ.'));
      }

      await Address.deleteOne({ _id: address._id });
      return res.status(200).json({ Success: true, Message: 'Đã xoá địa chỉ.' });
    } catch (error) {
      console.error('Delete address error:', error);
      return res.status(500).json(ApiResponse.error('Internal server error'));
    }
  });

  // PATCH /api/address/:id/set-default
  router.patch('/:id/set-default', async (req, res) => {
    try {
      const userId = getUserId(req);
      if (!userId) {
        return res.status(401).json(ApiResponse.error('Không xác định được người dùng.'));
      }

      const addresses = await Address.find({ userId });
      const target = addresses.find((a) => a._id.toString() === req.params.id);
      if (!target) {
        return res.status(404).json(ApiResponse.error('Không tìm thấy địa chỉ.'));
      }

      await Address.updateMany({ userId }, { $set: { isDefault: false } });
      target.isDefault = true;
      await target.save();

      return res.status(200).json({ Success: true, Data: target });
    } catch (error) {
      console.error('Set default address error:', error);
      return res.status(500).json(ApiResponse.error('Internal server error'));
    }
  });

  return router;
}
